usr/lib/common/key_mgr.c

```diff
@@ -194,46 +198,37 @@ CK_RV key_mgr_generate_key(STDLL_TokData_t *tokdata,
     // to their appropriate values.  this only applies to CKO_SECRET_KEY
     // and CKO_PRIVATE_KEY objects
     //
-    flag = template_attribute_find(key_obj->template, CKA_SENSITIVE, &attr);
-    if (flag == TRUE) {
-        flag = *(CK_BBOOL *) attr->pValue;
-
-        rc = build_attribute(CKA_ALWAYS_SENSITIVE, &flag, sizeof(CK_BBOOL),
-                             &new_attr);
-        if (rc != CKR_OK) {
-            TRACE_DEVEL("build attribute failed.\n");
-            goto error;
-        }
-        template_update_attribute(key_obj->template, new_attr);
-
-    } else {
-        rc = CKR_FUNCTION_FAILED;
+    rc = template_attribute_get_bool(key_obj->template, CKA_SENSITIVE, &flag);
+    if (rc != CKR_OK) {
         TRACE_ERROR("Failed to find CKA_SENSITIVE in key object template.\n");
         goto error;
     }
 
+    rc = build_attribute(CKA_ALWAYS_SENSITIVE, &flag, sizeof(CK_BBOOL),
+                         &new_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build attribute failed.\n");
+        goto error;
+    }
+    template_update_attribute(key_obj->template, new_attr);
```

Comment:
template_update_attribute  can only fail if one of its arguments is NULL, which is the callers responsibility. There are no other error cases. 
However, I agree that this is not really nice the way is is. I would also prefer to have return code checking there. Unfortunately, there are lots of places where template_update_attribute is used without error checking....

Reply:
Looking at the code of template_update_attribute, there is another way it can fail: memory error.  In that case, the function forgets about all attributes in the template and still returns success.  See bug #339 which might be more theoretic, but nevertheless I think we should fix that.

Reply:
Yes, agreed. We should fix template_update_attribute to return an error in case of a memory error (and not destroy the template). Once we have changed template_update_attribute we must also update all callers of template_update_attribute and add proper return code checking.
But this will be a separate commit in any case.

Reply:
I have just pushed a commit for fixing the template_update_attribute problem (https://github.com/opencryptoki/opencryptoki/issues/339) and also all its callers to check the return code:
https://github.com/opencryptoki/opencryptoki/pull/336/commits/136570f2708d89e1de47a659587b137319c02a0f
When you look at the size of commit you probably understand why I wanted to fix this in a separate commit... ;-)

Reply:
Nice work.  Sorry, there was a misunderstanding.  I did not want to suggest to fix this in this commit.  I just saw this and commented even though it has nothing to do with this commit...